### PR TITLE
dvcfs: open: raise FileNotFoundError on missing version info

### LINKED ISSUE #5692: metrics diff --show-md master: 'NoneType' object is not subscriptable
# Bug Report

Within master branch:
```sh
dvc metrics diff --show-md master 

ERROR: unexpected error - 'NoneType' object is not subscriptable      

Having any troubles? Hit us up at https://dvc.org/support, we are always happy to help!
```

This also happens with dvc plots
```sh
dvc plots diff --target loss.csv --show-vega master

ERROR: unexpected error - 'NoneType' object is not subscriptable

Having any troubles? Hit us up at https://dvc.org/support, we are always happy to help!
```

## Description

Doing that OP within master throws that error. This is undesired for CI/CD environments.
 
### Reproduce

```ssh
git clone https://github.com/DavidGOrtega/cml-dvc-basic.git
cd cml-dvc-basic
dvc init
dvc remote add -d myremote s3://your-remote
python get_data.py
dvc add data
```
```ssh
dvc push --run-cache

WARNING: Output 'metrics.json'(stage: 'mystage') is missing version info. Cache for it will not be collected. Use `dvc repro` to get your pipeline up to date.
WARNING: Output 'loss.csv'(stage: 'mystage') is missing version info. Cache for it will not be collected. Use `dvc repro` to get your pipeline up to date.
5 files pushed  
```

```ssh
git add --all
git commit -m 'ci ready' 
```

At this point should suffice for CML looking for doing repro within the CI but unfortunately it does not work 

```ssh
dvc repro --pull

'data.dvc' didn't change, skipping                                    
Running stage 'mystage':
> python train.py
/usr/local/lib/python2.7/dist-packages/sklearn/neural_network/multilayer_perceptron.py:562: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (23) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)
Generating lock file 'dvc.lock'                                                                                                    
Updating lock file 'dvc.lock'

To track the changes with git, run:

        git add .gitignore dvc.lock
Use `dvc push` to send your updates to remote storage.
```

```ssh
dvc metrics diff -v --show-md -v HEAD HEAD

2021-03-28 16:27:10,627 TRACE: Namespace(a_rev='HEAD', all=False, b_rev='HEAD', cd='.', cmd='diff', cprofile=False, cprofile_dump=None, func=<class 'dvc.command.metrics.CmdMetricsDiff'>, instrument=False, instrument_open=False, no_path=False, pdb=False, precision=None, quiet=0, recursive=False, show_json=False, show_md=True, targets=None, verbose=2, version=None)
2021-03-28 16:27:10,963 DEBUG: Check for update is enabled.
2021-03-28 16:27:11,069 TRACE: params.yaml does not exist, it won't be used in parametrization
2021-03-28 16:27:11,070 TRACE: Context during resolution of stage mystage:
{}
2021-03-28 16:27:11,414 TRACE: params.yaml does not exist, it won't be used in parametrization
2021-03-28 16:27:11,415 TRACE: Context during resolution of stage mystage:
{}
2021-03-28 16:27:11,416 DEBUG: Lockfile for 'dvc.yaml' not found
2021-03-28 16:27:11,582 ERROR: unexpected error - 'NoneType' object is not subscriptable
------------------------------------------------------------
Traceback (most recent call last):
  File "dvc/main.py", line 50, in main
  File "dvc/command/metrics.py", line 149, in run
  File "dvc/repo/metrics/__init__.py", line 13, in diff
  File "dvc/repo/metrics/diff.py", line 23, in diff
  File "dvc/repo/metrics/diff.py", line 8, in _get_metrics
  File "dvc/repo/metrics/__init__.py", line 8, in show
  File "dvc/repo/__init__.py", line 49, in wrapper
  File "dvc/repo/metrics/show.py", line 120, in show
  File "dvc/repo/metrics/show.py", line 82, in _read_metrics
  File "dvc/utils/serialize/_yaml.py", line 20, in load_yaml
  File "dvc/utils/serialize/_common.py", line 50, in _load_data
  File "dvc/fs/repo.py", line 143, in open
  File "dvc/fs/dvc.py", line 81, in open
  File "dvc/objects/db/base.py", line 68, in hash_to_path_info
TypeError: 'NoneType' object is not subscriptable
------------------------------------------------------------
2021-03-28 16:27:12,462 DEBUG: Version info for developers:
DVC version: 2.0.6 (deb)
---------------------------------
Platform: Python 3.7.10 on Linux-4.19.76-linuxkit-x86_64-with-debian-buster-sid
Supports: All remotes
Cache types: hardlink, symlink
Cache directory: fuse.grpcfuse on grpcfuse
Caches: local
Remotes: s3
Workspace directory: fuse.grpcfuse on grpcfuse
Repo: dvc, git

Having any troubles? Hit us up at https://dvc.org/support, we are always happy to help!
2021-03-28 16:27:12,473 DEBUG: Analytics is enabled.
2021-03-28 16:27:12,661 DEBUG: Trying to spawn '['daemon', '-q', 'analytics', '/tmp/tmpjo47u5rj']'
2021-03-28 16:27:12,667 DEBUG: Spawned '['daemon', '-q', 'analytics', '/tmp/tmpjo47u5rj']'
[18640] Failed to execute script __main__
```

plots does not work either
```ssh
dvc plots diff -v --target loss.csv --show-vega HEAD HEAD
2021-03-28 16:38:29,322 DEBUG: Check for update is enabled.
2021-03-28 16:38:29,442 ERROR: 'loss.csv' does not exist.
------------------------------------------------------------
Traceback (most recent call last):
  File "dvc/command/plots.py", line 37, in run
  File "dvc/command/plots.py", line 80, in _func
  File "dvc/repo/plots/__init__.py", line 173, in diff
  File "dvc/repo/plots/diff.py", line 16, in diff
  File "dvc/repo/plots/__init__.py", line 160, in show
dvc.exceptions.MetricDoesNotExistError: 'loss.csv' does not exist.
------------------------------------------------------------
2021-03-28 16:38:29,450 DEBUG: Analytics is enabled.
2021-03-28 16:38:29,506 DEBUG: Trying to spawn '['daemon', '-q', 'analytics', '/tmp/tmpjufpxbo2']'
2021-03-28 16:38:29,508 DEBUG: Spawned '['daemon', '-q', 'analytics', '/tmp/tmpjufpxbo2']'
root@97f8cc35d780:/iterative/cml-dvc-basic# [18685] Failed to execute script __main__
dvc plots diff -v --target loss.csv --show-vega HEAD
2021-03-28 16:38:47,913 DEBUG: Check for update is enabled.
2021-03-28 16:38:48,008 DEBUG: Lockfile for 'dvc.yaml' not found
2021-03-28 16:38:48,332 DEBUG: Lockfile for 'dvc.yaml' not found
2021-03-28 16:38:48,482 ERROR: unexpected error - 'NoneType' object is not subscriptable
------------------------------------------------------------
Traceback (most recent call last):
  File "dvc/main.py", line 50, in main
  File "dvc/command/plots.py", line 37, in run
  File "dvc/command/plots.py", line 80, in _func
  File "dvc/repo/plots/__init__.py", line 173, in diff
  File "dvc/repo/plots/diff.py", line 16, in diff
  File "dvc/repo/plots/__init__.py", line 149, in show
  File "dvc/repo/plots/__init__.py", line 86, in collect
  File "dvc/fs/repo.py", line 143, in open
  File "dvc/fs/dvc.py", line 81, in open
  File "dvc/objects/db/base.py", line 68, in hash_to_path_info
TypeError: 'NoneType' object is not subscriptable
------------------------------------------------------------
2021-03-28 16:38:49,413 DEBUG: Version info for developers:
DVC version: 2.0.6 (deb)
---------------------------------
Platform: Python 3.7.10 on Linux-4.19.76-linuxkit-x86_64-with-debian-buster-sid
Supports: All remotes
Cache types: hardlink, symlink
Cache directory: fuse.grpcfuse on grpcfuse
Caches: local
Remotes: s3
Workspace directory: fuse.grpcfuse on grpcfuse
Repo: dvc, git

Having any troubles? Hit us up at https://dvc.org/support, we are always happy to help!
2021-03-28 16:38:49,420 DEBUG: Analytics is enabled.
2021-03-28 16:38:49,597 DEBUG: Trying to spawn '['daemon', '-q', 'analytics', '/tmp/tmpiyh0i0y5']'
2021-03-28 16:38:49,601 DEBUG: Spawned '['daemon', '-q', 'analytics', '/tmp/tmpiyh0i0y5']'
root@97f8cc35d780:/iterative/cml-dvc-basic# [18702] Failed to execute script __main__
```

If we add dvc.lock diff will work

```ssh
git add dvc.lock
git commit -m 'dvc.lock'

dvc metrics diff -v --show-md -v HEAD HEAD
2021-03-28 16:27:48,548 TRACE: Namespace(a_rev='HEAD', all=False, b_rev='HEAD', cd='.', cmd='diff', cprofile=False, cprofile_dump=None, func=<class 'dvc.command.metrics.CmdMetricsDiff'>, instrument=False, instrument_open=False, no_path=False, pdb=False, precision=None, quiet=0, recursive=False, show_json=False, show_md=True, targets=None, verbose=2, version=None)
2021-03-28 16:27:48,864 DEBUG: Check for update is enabled.
2021-03-28 16:27:48,971 TRACE: params.yaml does not exist, it won't be used in parametrization
2021-03-28 16:27:48,972 TRACE: Context during resolution of stage mystage:
{}
2021-03-28 16:27:49,332 TRACE: params.yaml does not exist, it won't be used in parametrization
2021-03-28 16:27:49,333 TRACE: Context during resolution of stage mystage:
{}
2021-03-28 16:27:49,341 TRACE: Assuming '/iterative/cml-dvc-basic/.dvc/cache/be/3703304dfc59f95af156f8a354e68a' is unchanged since it is read-only
| Path   | Metric   | Old   | New   | Change   |
|--------|----------|-------|-------|----------|

2021-03-28 16:27:49,420 DEBUG: Analytics is enabled.
2021-03-28 16:27:49,507 DEBUG: Trying to spawn '['daemon', '-q', 'analytics', '/tmp/tmpdzwzsmo5']'
2021-03-28 16:27:49,511 DEBUG: Spawned '['daemon', '-q', 'analytics', '/tmp/tmpdzwzsmo5']'
[18675] Failed to execute script __main__
```

Additionally plots does not work with another error
```ssh
dvc plots diff -v --target loss.csv --show-vega HEAD HEAD
2021-03-28 16:43:30,479 DEBUG: Check for update is enabled.
2021-03-28 16:43:30,593 ERROR: 'loss.csv' does not exist.
------------------------------------------------------------
Traceback (most recent call last):
  File "dvc/command/plots.py", line 37, in run
  File "dvc/command/plots.py", line 80, in _func
  File "dvc/repo/plots/__init__.py", line 173, in diff
  File "dvc/repo/plots/diff.py", line 16, in diff
  File "dvc/repo/plots/__init__.py", line 160, in show
dvc.exceptions.MetricDoesNotExistError: 'loss.csv' does not exist.
------------------------------------------------------------
2021-03-28 16:43:30,608 DEBUG: Analytics is enabled.
2021-03-28 16:43:30,663 DEBUG: Trying to spawn '['daemon', '-q', 'analytics', '/tmp/tmp22ehhoy5']'
2021-03-28 16:43:30,665 DEBUG: Spawned '['daemon', '-q', 'analytics', '/tmp/tmp22ehhoy5']'
root@97f8cc35d780:/iterative/cml-dvc-basic# [18722] Failed to execute script __main__
```

### Expected

No diff message instead of that non descriptive error.

### Environment information

```console
$ dvc doctor

DVC version: 2.0.6 (deb)
---------------------------------
Platform: Python 3.7.10 on Linux-4.19.76-linuxkit-x86_64-with-debian-buster-sid
Supports: All remotes
Cache types: hardlink, symlink
Cache directory: fuse.grpcfuse on grpcfuse
Caches: local
Remotes: s3
Workspace directory: fuse.grpcfuse on grpcfuse
Repo: dvc, git
```

### PR DESCRIPTION
**Pull request by efiop**
Fixes #5692 

* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

File: tests/unit/fs/test_dvc.py [hunk 1]
```diff
@@ -39,6 +41,18 @@ def test_open_dirty_hash(tmp_dir, dvc):
...
+def test_open_no_remote(tmp_dir, dvc):
+    tmp_dir.dvc_gen("file", "file")
+    (tmp_dir / "file").unlink()
+    remove(dvc.odb.local.cache_dir)
+
+    fs = DvcFileSystem(repo=dvc)
+    with pytest.raises(FileNotFoundError) as exc_info:
+        with fs.open(PathInfo(tmp_dir) / "file", "r"):
+            pass
+    assert isinstance(exc_info.value.__cause__, NoRemoteError)
```
Comment (efiop) [inline tests/unit/fs/test_dvc.py:53]: Added since we were getting a false negative previously in dirty_no_hash

File: dvc/fs/dvc.py [hunk 2]
```diff
@@ -68,13 +68,17 @@ def open(  # type: ignore
             raise IsADirectoryError
 
         out = outs[0]
+
+        if not out.hash_info:
+            raise FileNotFoundError
+
```
Comment (skshetry) [inline dvc/fs/dvc.py:74]: Is this covered already? The tests below seem to be for another change.

**Review by skshetry — APPROVED** [review]
Looks good to me. Just a small comment inlined.

File: dvc/fs/dvc.py (hunk 2 above)
Comment (efiop) [reply dvc/fs/dvc.py:74]: Yeah, there is an existing tests below that tests for this and I wrote a new one for NoRemoteError.

**Pull request merged**